File: src/mixi_backup.ts
/**
 * mixi_backup.ts
 *
 * Back up your mixi diary entries, photos, and comments.
 *
 * Usage (after `npm run build`):
 *   node dist/mixi_backup.js --email you@example.com --password yourpass [--output ./backup]
 */

import { load } from "cheerio";
import type { CheerioAPI } from "cheerio";
import * as fs from "fs";
import * as path from "path";
import { Command } from "commander";

const BASE_URL = "https://mixi.jp";
const LOGIN_URL = `${BASE_URL}/login.pl`;
const DIARY_LIST_URL = `${BASE_URL}/list_diary.pl`;
const VIEW_DIARY_URL = `${BASE_URL}/view_diary.pl`;

/** Milliseconds to wait between requests to avoid hammering the server. */
const REQUEST_INTERVAL_MS = 1500;

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface DiaryEntry {
  id: string;
  ownerId: string;
  title: string;
  url: string;
  date?: string;
  bodyHtml?: string;
  bodyText?: string;
  photoUrls?: string[];
  localPhotos?: string[];
  comments?: Comment[];
}

export interface Comment {
  author: string;
  body: string;
  date: string;
}

// ---------------------------------------------------------------------------
// Cookie jar — thin wrapper around the built-in fetch
// ---------------------------------------------------------------------------

/**
 * Minimal cookie jar that stores cookies and injects them into every request.
 * This replaces axios / node-fetch so we only use the Node 20 built-in fetch.
 */
export class CookieJar {
  private cookies: Map<string, string> = new Map();

  /** Parse Set-Cookie headers from a Response and store the values. */
  ingest(response: Response): void {
    const raw = response.headers.getSetCookie?.() ?? [];
    for (const cookie of raw) {
      const pair = cookie.split(";")[0].trim();
      const eqIdx = pair.indexOf("=");
      if (eqIdx === -1) continue;
      const name = pair.slice(0, eqIdx).trim();
      const value = pair.slice(eqIdx + 1).trim();
      this.cookies.set(name, value);
    }
  }

  /** Return a Cookie header value for the current jar. */
  header(): string {
    return Array.from(this.cookies.entries())
      .map(([k, v]) => `${k}=${v}`)
      .join("; ");
  }
}

// ---------------------------------------------------------------------------
// HTTP helpers
// ---------------------------------------------------------------------------

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Perform a fetch request, automatically injecting the session cookies and
 * storing any new cookies from the response.
 *
 * An optional `fetchFn` can be injected (e.g. in tests) instead of the
 * global `fetch`.
 */
export async function sessionFetch(
  jar: CookieJar,
  url: string,
  options: RequestInit = {},
  fetchFn: FetchFn = fetch
): Promise<Response> {
  const headers: Record<string, string> = {
    "User-Agent":
      "Mozilla/5.0 (Windows NT 10.0; Win64; x64) " +
      "AppleWebKit/537.36 (KHTML, like Gecko) " +
      "Chrome/124.0.0.0 Safari/537.36",
    Cookie: jar.header(),
    ...(options.headers as Record<string, string>),
  };

  const response = await fetchFn(url, { ...options, headers });
  jar.ingest(response);

  if (!response.ok) {
    throw new Error(`HTTP ${response.status} ${response.statusText} — ${url}`);
  }
  return response;
}

// ---------------------------------------------------------------------------
// HTML parsing helpers
// ---------------------------------------------------------------------------

export function extractOwnerId(html: string, responseUrl: string): string | null {
  // 1. Try query-string of the response URL
  try {
    const u = new URL(responseUrl);
    const id = u.searchParams.get("owner_id");
    if (id) return id;
  } catch {
    // ignore
  }

  // 2. Scrape a profile link
  const $ = load(html);
  const profileLink = $("a[href*='view_profile.pl']").first().attr("href") ?? "";
  const m = profileLink.match(/[?&]id=(\d+)/);
  if (m) return m[1];

  // 3. Hidden input
  const hidden = $("input[name='owner_id']").first().val();
  if (hidden) return String(hidden);

  return null;
}

export function parseDiaryListPage(
  html: string,
  fallbackOwnerId: string
): DiaryEntry[] {
  const $ = load(html);
  const seen = new Set<string>();
  const entries: DiaryEntry[] = [];

  $("a[href*='view_diary.pl']").each((_i, el) => {
    const href = $(el).attr("href") ?? "";
    const mId = href.match(/[?&]id=(\d+)/);
    const mOwner = href.match(/[?&]owner_id=(\d+)/);
    if (!mId) return;
    const id = mId[1];
    if (seen.has(id)) return;
    seen.add(id);
    entries.push({
      id,
      ownerId: mOwner ? mOwner[1] : fallbackOwnerId,
      title: $(el).text().trim(),
      url: href.startsWith("http") ? href : `${BASE_URL}${href}`,
    });
  });

  return entries;
}

export function hasNextPage(html: string): boolean {
  const $ = load(html);
  return $("a").filter((_i, el) => /次|next/i.test($(el).text())).length > 0;
}

export function extractDiaryDetail(
  html: string,
  entry: DiaryEntry
): DiaryEntry {
  const $ = load(html);

  // Body
  const bodyEl =
    $("[class*='diaryBody'], [class*='diary-body'], [class*='contents']").first();
  const bodyHtml = bodyEl.html() ?? "";
  const bodyText = bodyEl.text().trim();

  // Date
  const dateEl = $("[class*='date'], [class*='time'], [class*='timestamp']").first();
  const date = dateEl.text().trim();

  // Photos
  const photoUrls: string[] = [];
  $("img").each((_i, el) => {
    const src = $(el).attr("src") ?? "";
    if (src.startsWith("http") && /photo|img|image/i.test(src)) {
      if (!photoUrls.includes(src)) photoUrls.push(src);
    }
  });

  // Comments
  const comments = extractComments($);

  return {
    ...entry,
    date,
    bodyHtml,
    bodyText,
    photoUrls,
    comments,
  };
}

function extractComments($: CheerioAPI): Comment[] {
  const comments: Comment[] = [];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  $("[class*='comment'], [class*='Comment']").each((_i: number, item: any) => {
    const author =
      $(item).find("[class*='name'], [class*='author'], [class*='nick']").first().text().trim();
    const body =
      $(item).find("[class*='body'], [class*='text'], [class*='content']").first().text().trim();
    const date =
      $(item).find("[class*='date'], [class*='time']").first().text().trim();
    if (author || body) {
      comments.push({ author, body, date });
    }
  });
  return comments;
}

/** Minimal fetch signature used internally so we avoid the overly broad built-in. */
type FetchFn = (url: string, init?: RequestInit) => Promise<Response>;

export class MixiBackup {
  private jar = new CookieJar();
  private ownerId: string | null = null;

  constructor(private outputDir: string = "./backup") {}

  // ── Auth ────────────────────────────────────────────────────────────────

  async login(email: string, password: string): Promise<void> {
    console.log(`Logging in as ${email} …`);
    const body = new URLSearchParams({
      next_url: "/home.pl",
      email,
      password,
    });

    const response = await sessionFetch(this.jar, LOGIN_URL, {
      method: "POST",
      body,
      redirect: "follow",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
    });

    const html = await response.text();
    const finalUrl = response.url;

    if (!finalUrl.includes("home.pl") && !finalUrl.includes("mixi_id")) {
      throw new Error("ログインに失敗しました。メールアドレスとパスワードを確認してください。");
    }

    this.ownerId = extractOwnerId(html, finalUrl);
    if (!this.ownerId) {
      throw new Error("ログイン後にowner_idを取得できませんでした。");
    }
    console.log(`ログイン成功 (owner_id=${this.ownerId})`);
  }

  // ── Diary list ──────────────────────────────────────────────────────────

  async fetchDiaryList(): Promise<DiaryEntry[]> {
    const entries: DiaryEntry[] = [];
    let page = 1;

    while (true) {
      console.log(`日記一覧を取得中 (page ${page}) …`);
      const url = `${DIARY_LIST_URL}?page=${page}&owner_id=${this.ownerId}`;
      const response = await sessionFetch(this.jar, url);
      const html = await response.text();

      const newEntries = parseDiaryListPage(html, this.ownerId!);
      if (newEntries.length === 0) break;
      entries.push(...newEntries);

      if (!hasNextPage(html)) break;
      page++;
      await sleep(REQUEST_INTERVAL_MS);
    }

    console.log(`${entries.length} 件の日記が見つかりました。`);
    return entries;
  }

  // ── Individual entry ────────────────────────────────────────────────────

  async fetchDiaryEntry(entry: DiaryEntry): Promise<DiaryEntry> {
    console.log(`日記を取得中 id=${entry.id}: ${entry.title}`);
    const url = `${VIEW_DIARY_URL}?id=${entry.id}&owner_id=${entry.ownerId}`;
    const response = await sessionFetch(this.jar, url);
    const html = await response.text();
    return extractDiaryDetail(html, entry);
  }

  // ── Photos ──────────────────────────────────────────────────────────────

  async downloadPhoto(
    photoUrl: string,
    destDir: string,
    fetchFn: FetchFn = (u, o) => sessionFetch(this.jar, u, o)
  ): Promise<string | null> {
    try {
      const filename = path.basename(new URL(photoUrl).pathname) || "photo.jpg";
      const destPath = path.join(destDir, filename);
      if (fs.existsSync(destPath)) return filename;

      const response = await fetchFn(photoUrl, {});
      if (!response.ok) throw new Error(`HTTP ${response.status}`);
      const buffer = Buffer.from(await response.arrayBuffer());
      fs.writeFileSync(destPath, buffer);
      return filename;
    } catch (err) {
      console.warn(`写真のダウンロードに失敗しました: ${photoUrl}`, err);
      return null;
    }
  }

  // ── Persistence ─────────────────────────────────────────────────────────

  private diaryDir(entryId: string): string {
    const dir = path.join(this.outputDir, "diaries", entryId);
    fs.mkdirSync(dir, { recursive: true });
    return dir;
  }

  saveEntry(entry: DiaryEntry): void {
    const dir = this.diaryDir(entry.id);

    if (entry.bodyHtml) {
      fs.writeFileSync(path.join(dir, "body.html"), entry.bodyHtml, "utf-8");
    }

    // Exclude the raw HTML blob from the JSON to keep it readable
    const { bodyHtml: _bodyHtml, ...meta } = entry;
    fs.writeFileSync(
      path.join(dir, "entry.json"),
      JSON.stringify(meta, null, 2),
      "utf-8"
    );
  }

  // ── Orchestration ───────────────────────────────────────────────────────

  async run(email: string, password: string): Promise<void> {
    fs.mkdirSync(this.outputDir, { recursive: true });

    await this.login(email, password);
    const entries = await this.fetchDiaryList();

    const index: Omit<DiaryEntry, "bodyHtml">[] = [];

    for (const entry of entries) {
      let full: DiaryEntry;
      try {
        full = await this.fetchDiaryEntry(entry);
      } catch (err) {
        console.warn(`日記 ${entry.id} のスキップ:`, err);
        continue;
      }

      // Download photos
      if (full.photoUrls && full.photoUrls.length > 0) {
        const photoDir = this.diaryDir(full.id);
        const localPhotos: string[] = [];
        for (const photoUrl of full.photoUrls) {
          const fname = await this.downloadPhoto(photoUrl, photoDir);
          if (fname) localPhotos.push(fname);
          await sleep(REQUEST_INTERVAL_MS);
        }
        full.localPhotos = localPhotos;
      }

      this.saveEntry(full);
      const { bodyHtml: _bodyHtml, ...meta } = full;
      index.push(meta);
      await sleep(REQUEST_INTERVAL_MS);
    }

    fs.writeFileSync(
      path.join(this.outputDir, "index.json"),
      JSON.stringify(index, null, 2),
      "utf-8"
    );
    console.log(`バックアップ完了: ${index.length} 件を ${this.outputDir} に保存しました。`);
  }
}

// ---------------------------------------------------------------------------
// CLI entry-point
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  const program = new Command();
  program
    .name("mixi-backup")
    .description("mixiの日記・写真・コメントをバックアップします。")
    .requiredOption("--email <email>", "mixiのメールアドレス")
    .requiredOption("--password <password>", "mixiのパスワード")
    .option("--output <dir>", "保存先ディレクトリ", "./backup")
    .parse(process.argv);

  const opts = program.opts<{ email: string; password: string; output: string }>();
  const backup = new MixiBackup(opts.output);
  await backup.run(opts.email, opts.password);
}

if (require.main === module) {
  main().catch((err) => {
    console.error(err instanceof Error ? err.message : err);
    process.exit(1);
  });
}

File: mixi_backup.py
"""mixi_backup.py - Back up your mixi diary entries, photos, and comments.

Usage:
    python mixi_backup.py --email you@example.com --password yourpass [--output ./backup]
"""

import argparse
import json
import logging
import os
import re
import time
from urllib.parse import urljoin, urlparse, parse_qs

import requests
from bs4 import BeautifulSoup

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(message)s",
)
logger = logging.getLogger(__name__)

BASE_URL = "https://mixi.jp"
LOGIN_URL = f"{BASE_URL}/login.pl"
DIARY_LIST_URL = f"{BASE_URL}/list_diary.pl"
VIEW_DIARY_URL = f"{BASE_URL}/view_diary.pl"

REQUEST_INTERVAL = 1.5  # seconds between requests to be polite


class MixiBackup:
    """Downloads all diary entries, comments and photos from a mixi account."""

    def __init__(self, output_dir: str = "./backup") -> None:
        self.output_dir = output_dir
        self.session = requests.Session()
        self.session.headers.update(
            {
                "User-Agent": (
                    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                    "AppleWebKit/537.36 (KHTML, like Gecko) "
                    "Chrome/124.0.0.0 Safari/537.36"
                )
            }
        )
        self.owner_id: str | None = None

    # ------------------------------------------------------------------
    # Authentication
    # ------------------------------------------------------------------

    def login(self, email: str, password: str) -> None:
        """Log in to mixi and store the session cookies."""
        logger.info("Logging in as %s …", email)
        payload = {
            "next_url": "/home.pl",
            "email": email,
            "password": password,
        }
        response = self.session.post(LOGIN_URL, data=payload, allow_redirects=True)
        response.raise_for_status()

        if "home.pl" not in response.url and "mixi_id" not in response.url:
            raise RuntimeError(
                "Login failed — please check your e-mail and password."
            )

        self.owner_id = self._extract_owner_id(response)
        if not self.owner_id:
            raise RuntimeError("Could not determine your mixi owner ID after login.")
        logger.info("Logged in successfully (owner_id=%s).", self.owner_id)

    def _extract_owner_id(self, response: requests.Response) -> str | None:
        """Try to extract the owner_id from the post-login page."""
        # Check URL first (mixi sometimes redirects to /home.pl?...)
        parsed = urlparse(response.url)
        qs = parse_qs(parsed.query)
        if "owner_id" in qs:
            return qs["owner_id"][0]

        # Fall back to scraping the page
        soup = BeautifulSoup(response.text, "lxml")

        # Links like /view_profile.pl?id=12345
        profile_link = soup.find("a", href=re.compile(r"view_profile\.pl\?id=(\d+)"))
        if profile_link:
            m = re.search(r"id=(\d+)", profile_link["href"])
            if m:
                return m.group(1)

        # Hidden input on the page
        hidden = soup.find("input", {"name": "owner_id"})
        if hidden and hidden.get("value"):
            return hidden["value"]

        return None

    # ------------------------------------------------------------------
    # Diary listing
    # ------------------------------------------------------------------

    def fetch_diary_list(self) -> list[dict]:
        """Return a list of dicts with ``id``, ``owner_id``, ``title``, ``date``."""
        entries: list[dict] = []
        page = 1

        while True:
            logger.info("Fetching diary list page %d …", page)
            params = {"page": page, "owner_id": self.owner_id}
            response = self.session.get(DIARY_LIST_URL, params=params)
            response.raise_for_status()
            soup = BeautifulSoup(response.text, "lxml")

            new_entries = self._parse_diary_list_page(soup)
            if not new_entries:
                break
            entries.extend(new_entries)

            # Check whether a "next page" link exists
            if not self._has_next_page(soup):
                break
            page += 1
            time.sleep(REQUEST_INTERVAL)

        logger.info("Found %d diary entries.", len(entries))
        return entries

    def _parse_diary_list_page(self, soup: BeautifulSoup) -> list[dict]:
        entries = []
        # mixi diary list items are in <li> tags inside a <ul class="list-...">
        for link in soup.find_all("a", href=re.compile(r"view_diary\.pl")):
            href = link.get("href", "")
            m_id = re.search(r"id=(\d+)", href)
            m_owner = re.search(r"owner_id=(\d+)", href)
            if not m_id:
                continue
            entry = {
                "id": m_id.group(1),
                "owner_id": m_owner.group(1) if m_owner else self.owner_id,
                "title": link.get_text(strip=True),
                "url": urljoin(BASE_URL, href),
            }
            # Avoid duplicates
            if not any(e["id"] == entry["id"] for e in entries):
                entries.append(entry)
        return entries

    def _has_next_page(self, soup: BeautifulSoup) -> bool:
        next_link = soup.find("a", string=re.compile(r"次|next"))
        return next_link is not None

    # ------------------------------------------------------------------
    # Individual diary entry
    # ------------------------------------------------------------------

    def fetch_diary_entry(self, entry: dict) -> dict:
        """Fetch the full content (body, comments, photo URLs) for one entry."""
        logger.info("Fetching diary %s: %s", entry["id"], entry.get("title", ""))
        params = {"id": entry["id"], "owner_id": entry["owner_id"]}
        response = self.session.get(VIEW_DIARY_URL, params=params)
        response.raise_for_status()
        soup = BeautifulSoup(response.text, "lxml")

        body_html, body_text = self._extract_body(soup)
        date = self._extract_date(soup)
        photos = self._extract_photo_urls(soup)
        comments = self._extract_comments(soup)

        return {
            **entry,
            "date": date,
            "body_html": body_html,
            "body_text": body_text,
            "photo_urls": photos,
            "comments": comments,
        }

    def _extract_body(self, soup: BeautifulSoup) -> tuple[str, str]:
        """Return (raw_html, plain_text) of the diary body."""
        # mixi wraps diary body in a <div class="diaryBody"> or similar
        body_div = (
            soup.find("div", class_=re.compile(r"diaryBody|diary-body|contents"))
            or soup.find("div", id=re.compile(r"diaryBody|diary_body"))
        )
        if body_div:
            return str(body_div), body_div.get_text(separator="\n", strip=True)
        return "", ""

    def _extract_date(self, soup: BeautifulSoup) -> str:
        date_el = soup.find(
            ["span", "div", "time"],
            class_=re.compile(r"date|time|timestamp", re.IGNORECASE),
        )
        if date_el:
            return date_el.get_text(strip=True)
        return ""

    def _extract_photo_urls(self, soup: BeautifulSoup) -> list[str]:
        urls = []
        for img in soup.find_all("img", src=re.compile(r"photo|img|image", re.IGNORECASE)):
            src = img.get("src", "")
            if src and src.startswith("http"):
                urls.append(src)
        return list(dict.fromkeys(urls))  # deduplicate, preserve order

    def _extract_comments(self, soup: BeautifulSoup) -> list[dict]:
        comments = []
        comment_items = soup.find_all(
            ["div", "li"],
            class_=re.compile(r"comment|Comment", re.IGNORECASE),
        )
        for item in comment_items:
            author_el = item.find(
                ["span", "div", "a"],
                class_=re.compile(r"name|author|nick", re.IGNORECASE),
            )
            body_el = item.find(
                ["span", "div", "p"],
                class_=re.compile(r"body|text|content", re.IGNORECASE),
            )
            date_el = item.find(
                ["span", "div"],
                class_=re.compile(r"date|time", re.IGNORECASE),
            )
            comment = {
                "author": author_el.get_text(strip=True) if author_el else "",
                "body": body_el.get_text(separator="\n", strip=True) if body_el else "",
                "date": date_el.get_text(strip=True) if date_el else "",
            }
            if comment["body"] or comment["author"]:
                comments.append(comment)
        return comments

    # ------------------------------------------------------------------
    # Photo download
    # ------------------------------------------------------------------

    def download_photo(self, url: str, dest_dir: str) -> str | None:
        """Download a photo to *dest_dir* and return the local filename."""
        try:
            filename = os.path.basename(urlparse(url).path) or "photo.jpg"
            dest_path = os.path.join(dest_dir, filename)
            if os.path.exists(dest_path):
                return filename
            resp = self.session.get(url, timeout=30)
            resp.raise_for_status()
            with open(dest_path, "wb") as f:
                f.write(resp.content)
            return filename
        except requests.RequestException as exc:
            logger.warning("Failed to download %s: %s", url, exc)
            return None

    # ------------------------------------------------------------------
    # Persistence helpers
    # ------------------------------------------------------------------

    def _diary_dir(self, entry_id: str) -> str:
        path = os.path.join(self.output_dir, "diaries", entry_id)
        os.makedirs(path, exist_ok=True)
        return path

    def save_entry(self, entry: dict) -> None:
        """Save the full entry data (JSON + HTML body) to disk."""
        d = self._diary_dir(entry["id"])

        # Save raw HTML body
        if entry.get("body_html"):
            with open(os.path.join(d, "body.html"), "w", encoding="utf-8") as f:
                f.write(entry["body_html"])

        # Save full metadata as JSON (exclude large html blob)
        meta = {k: v for k, v in entry.items() if k != "body_html"}
        with open(os.path.join(d, "entry.json"), "w", encoding="utf-8") as f:
            json.dump(meta, f, ensure_ascii=False, indent=2)

    # ------------------------------------------------------------------
    # Orchestration
    # ------------------------------------------------------------------

    def run(self, email: str, password: str) -> None:
        """Full backup: login → list → fetch each entry → download photos."""
        os.makedirs(self.output_dir, exist_ok=True)

        self.login(email, password)

        entries = self.fetch_diary_list()

        all_entries = []
        for entry in entries:
            try:
                full = self.fetch_diary_entry(entry)
            except requests.RequestException as exc:
                logger.warning("Skipping entry %s: %s", entry.get("id"), exc)
                continue

            # Download photos
            if full.get("photo_urls"):
                photo_dir = self._diary_dir(full["id"])
                local_photos = []
                for url in full["photo_urls"]:
                    fname = self.download_photo(url, photo_dir)
                    if fname:
                        local_photos.append(fname)
                    time.sleep(REQUEST_INTERVAL)
                full["local_photos"] = local_photos

            self.save_entry(full)
            all_entries.append({k: v for k, v in full.items() if k != "body_html"})
            time.sleep(REQUEST_INTERVAL)

        # Write summary index
        index_path = os.path.join(self.output_dir, "index.json")
        with open(index_path, "w", encoding="utf-8") as f:
            json.dump(all_entries, f, ensure_ascii=False, indent=2)
        logger.info("Backup complete. %d entries saved to %s", len(all_entries), self.output_dir)


# ---------------------------------------------------------------------------
# CLI entry-point
# ---------------------------------------------------------------------------

def main() -> None:
    parser = argparse.ArgumentParser(
        description="Back up your mixi diary entries, photos, and comments."
    )
    parser.add_argument("--email", required=True, help="Your mixi e-mail address")
    parser.add_argument("--password", required=True, help="Your mixi password")
    parser.add_argument(
        "--output",
        default="./backup",
        help="Directory to store the backup (default: ./backup)",
    )
    args = parser.parse_args()

    backup = MixiBackup(output_dir=args.output)
    backup.run(email=args.email, password=args.password)


if __name__ == "__main__":
    main()
